src/main/java/com/linecorp/armeria/common/logging/AbstractMessageLog.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.util.UnitFormatter;
+
+abstract class AbstractMessageLog<T extends MessageLog>
+        extends CompletableFuture<T> implements MessageLog, MessageLogBuilder {
+
+    private boolean startTimeNanosSet;
+    private long startTimeNanos;
+    private long contentLength;
+    private long endTimeNanos;
+    private List<Object> attachments;
+    private Throwable cause;
+
+    boolean start0() {
+        if (isDone()) {
+            return false;
+        }
+
+        if (!startTimeNanosSet) {
+            startTimeNanos = System.nanoTime();
+            startTimeNanosSet = true;
+        }
+
+        return true;
+    }
+
+    @Override
+    public long startTimeNanos() {
+        return startTimeNanos;
```

Comment:
Yeah, let me do some sanity check when `end()` is called. Note that a user will never have access to `startTimeNanos` until the `CompletableFuture` is fulfilled by `end()`.